zookeeper-api/src/main/java/org/apache/helix/zookeeper/api/client/RealmAwareZkClient.java

```diff
@@ -373,6 +379,91 @@ public String getZkRealmShardingKey() {
     public int getSessionTimeout() {
       return _sessionTimeout;
     }
+
+    public String getMsdsEndpoint() {
+      return _msdsEndpoint;
+    }
+
+    public HelixZkClient.ZkConnectionConfig createZkConnectionConfig()
+        throws IOException, InvalidRoutingDataException {
+      // Convert to a single-realm HelixZkClient's ZkConnectionConfig
+      if (_zkRealmShardingKey == null || _zkRealmShardingKey.isEmpty()) {
+        throw new ZkClientException(
+            "Cannot create ZkConnectionConfig because ZK realm sharding key is either null or empty!");
+      }
+
+      String zkAddress;
+      // Look up the ZK address for the given ZK realm sharding key
+      if (_msdsEndpoint == null || _msdsEndpoint.isEmpty()) {
+        zkAddress = HttpRoutingDataReader.getMetadataStoreRoutingData()
+            .getMetadataStoreRealm(_zkRealmShardingKey);
+      } else {
+        zkAddress = HttpRoutingDataReader.getMetadataStoreRoutingData(_msdsEndpoint)
+            .getMetadataStoreRealm(_zkRealmShardingKey);
+      }
+
+      return new HelixZkClient.ZkConnectionConfig(zkAddress).setSessionTimeout(_sessionTimeout);
+    }
+
+    public static class Builder {
+      private RealmMode _realmMode;
+      private String _zkRealmShardingKey;
+      private String _msdsEndpoint;
+      private int _sessionTimeout = DEFAULT_SESSION_TIMEOUT;
+
+      public Builder() {
+      }
+
+      public Builder setRealmMode(RealmMode mode) {
+        _realmMode = mode;
+        return this;
+      }
+
+      public Builder setZkRealmShardingKey(String shardingKey) {
+        _zkRealmShardingKey = shardingKey;
+        return this;
+      }
+
+      public Builder setMsdsEndpoint(String msdsEndpoint) {
+        _msdsEndpoint = msdsEndpoint;
+        return this;
+      }
+
+      public Builder setSessionTimeout(int sessionTimeout) {
+        _sessionTimeout = sessionTimeout;
+        return this;
+      }
+
+      public RealmAwareZkConnectionConfig build() {
+        validate();
+        return new RealmAwareZkConnectionConfig(this);
+      }
+
+      /**
+       * Validate the internal fields of the builder before creating an instance.
+       */
+      private void validate() {
+        boolean isRealmModeSet = _realmMode != null;
+        boolean isShardingKeySet = _zkRealmShardingKey != null && !_zkRealmShardingKey.isEmpty();
+        switch (isRealmModeSet ? _realmMode : RealmMode.MULTI_REALM) {
```

Comment:
It seems by default you set realmMode as multi realm. Why not set it multi realm by default in constructor ?
And if sharding key is set only, are we going to just allow it to have single realm by default?
 It seems easier to use and to check the code logic?

By default:
1. No sharding key or realm mode set, by default it is multi realm
2. Only sharding key is set, single realm
3. Sharding key and multi realm set, illegal

Reply:
That's what this piece of logic is doing.

Moving validation logic to a validate() function is a lot cleaner and promotes better organization. The constructor only constructs.

I am not understanding your question/comment clearly - could you clarify?

Reply:
This validation logic needs some time to actually understand. At a first glance, I was not able to understand the logic.

- if only sharding key is set and realm mode is not set, I would expect to create a single mode client, but this validation does not pass. Or is realm mode forced to set?

Reply:
I simplified the validation logic. Let me know if you think it's clearer.